File: 0907-sum-of-subarray-minimums/0907-sum-of-subarray-minimums.py
class Solution:
    def prev_small_index(self,arr,n):
        ans = [0] * n
        stack = []
        for i in range(n):
            while( stack and arr[stack[-1]] > arr[i]):
                stack.pop()
            ans[i] = -1 if not stack else stack[-1]
            stack.append(i)
        return ans
    def next_small_index(self,arr,n):
            ans = [0]*n
            stack = []
            for i in reversed(range(n)):
                while( stack and arr[stack[-1]] >= arr[i]):
                    stack.pop()
                ans[i] = n if not stack else stack[-1]
                stack.append(i)
            return ans
    def sumSubarrayMins(self, arr: List[int]) -> int:
        MOD = int(1e9 + 7) 
        n = len(arr)
        prev = self.prev_small_index(arr,n)
        next_ = self.next_small_index(arr,n)

        total = 0
        for i in range(n):
            left = i - prev[i]
            right = next_[i] - i
            total = (total + (arr[i] * left * right) % MOD) % MOD
        return total

File: 0907-sum-of-subarray-minimums/0907-sum-of-subarray-minimums.java
import java.util.*;

class Solution {

    private int[] prevSmallIndex(int[] arr, int n) {
        int[] ans = new int[n];
        Stack<Integer> stack = new Stack<>();

        for (int i = 0; i < n; i++) {
            while (!stack.isEmpty() && arr[stack.peek()] > arr[i]) {
                stack.pop();
            }
            ans[i] = stack.isEmpty() ? -1 : stack.peek();
            stack.push(i);
        }
        return ans;
    }

    private int[] nextSmallIndex(int[] arr, int n) {
        int[] ans = new int[n];
        Stack<Integer> stack = new Stack<>();

        for (int i = n - 1; i >= 0; i--) {
            while (!stack.isEmpty() && arr[stack.peek()] >= arr[i]) {
                stack.pop();
            }
            ans[i] = stack.isEmpty() ? n : stack.peek();
            stack.push(i);
        }
        return ans;
    }

    public int sumSubarrayMins(int[] arr) {
        int MOD = 1_000_000_007;
        int n = arr.length;

        int[] prev = prevSmallIndex(arr, n);
        int[] next = nextSmallIndex(arr, n);

        long total = 0;
        for (int i = 0; i < n; i++) {
            long left = i - prev[i];
            long right = next[i] - i;
            total = ( total + arr[i] * left * right) % MOD;
        }

        return (int) total;
    }
}
